# Telecom Customer Churn Analysis

## 01: Data Audit and Cleaning

### Objective
This notebook focuses on assessing the quality and structure of the Telecom Customer Churn dataframe before performiing exploratory analysis.

The data preparation process includes:
- Reviewing the dataset structure and data types
- Correcting inappropriate data types
- Checking for duplicate records
- Investigating missing values and determining whether they require treatment
- Validating categorical values for inconsistencies
- Checking numerical values for invalid or unreasonable observations
- Preparing a clean dataset for subsequent analysis

The cleaned dataset produced from this notebook will serve as the foundation for succeeding exploratory analysis notebooks.

In [1]:
import pandas as pd

In [2]:
# Load csv
churn_df = pd.read_csv('../data/raw/telecom_customer_churn.csv')

In [3]:
# Customize display options of DataFrame for better viewing in IDE
pd.set_option('display.max_columns', None, 'display.max_colwidth', None)

## Initial Data Inspection

Inspect the dataset structure, dimensions, column, data types, and missing values before making any transformations.

In [4]:
churn_df.head()

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason
0,0002-ORFBO,Female,37,Yes,0,Frazier Park,93225,34.827662,-118.999073,2,9,NaN,Yes,42.39,No,Yes,Cable,16.0,No,Yes,No,Yes,Yes,No,No,Yes,One Year,Yes,Credit Card,65.6,593.30,0.00,0,381.51,974.81,Stayed,NaN,NaN
1,0003-MKNFE,Male,46,No,0,Glendale,91206,34.162515,-118.203869,0,9,NaN,Yes,10.69,Yes,Yes,Cable,10.0,No,No,No,No,No,Yes,Yes,No,Month-to-Month,No,Credit Card,-4.0,542.40,38.33,10,96.21,610.28,Stayed,NaN,NaN
2,0004-TLHLJ,Male,50,No,0,Costa Mesa,92627,33.645672,-117.922613,0,4,Offer E,Yes,33.65,No,Yes,Fiber Optic,30.0,No,No,Yes,No,No,No,No,Yes,Month-to-Month,Yes,Bank Withdrawal,73.9,280.85,0.00,0,134.60,415.45,Churned,Competitor,Competitor had better devices
3,0011-IGKFF,Male,78,Yes,0,Martinez,94553,38.014457,-122.115432,1,13,Offer D,Yes,27.82,No,Yes,Fiber Optic,4.0,No,Yes,Yes,No,Yes,Yes,No,Yes,Month-to-Month,Yes,Bank Withdrawal,98.0,1237.85,0.00,0,361.66,1599.51,Churned,Dissatisfaction,Product dissatisfaction
4,0013-EXCHZ,Female,75,Yes,0,Camarillo,93010,34.227846,-119.079903,3,3,NaN,Yes,7.38,No,Yes,Fiber Optic,11.0,No,No,No,Yes,Yes,No,No,Yes,Month-to-Month,Yes,Credit Card,83.9,267.40,0.00,0,22.14,289.54,Churned,Dissatisfaction,Network reliability


In [5]:
churn_df.shape

(7043, 38)

In [6]:
churn_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Customer ID                        7043 non-null   str    
 1   Gender                             7043 non-null   str    
 2   Age                                7043 non-null   int64  
 3   Married                            7043 non-null   str    
 4   Number of Dependents               7043 non-null   int64  
 5   City                               7043 non-null   str    
 6   Zip Code                           7043 non-null   int64  
 7   Latitude                           7043 non-null   float64
 8   Longitude                          7043 non-null   float64
 9   Number of Referrals                7043 non-null   int64  
 10  Tenure in Months                   7043 non-null   int64  
 11  Offer                              3166 non-null   str    
 12  Pho

In [7]:
churn_df.describe()

,Age,Number of Dependents,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Avg Monthly Long Distance Charges,Avg Monthly GB Download,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue
count,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,6361.000000,5517.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000
mean,46.509726,0.468692,93486.070567,36.197455,-119.756684,1.951867,32.386767,25.420517,26.189958,63.596131,2280.381264,1.962182,6.860713,749.099262,3034.379056
std,16.750352,0.962802,1856.767505,2.468929,2.154425,3.001199,24.542061,14.200374,19.586585,31.204743,2266.220462,7.902614,25.104978,846.660055,2865.204542
min,19.000000,0.000000,90001.000000,32.555828,-124.301372,0.000000,1.000000,1.010000,2.000000,-10.000000,18.800000,0.000000,0.000000,0.000000,21.360000
25%,32.000000,0.000000,92101.000000,33.990646,-121.788090,0.000000,9.000000,13.050000,13.000000,30.400000,400.150000,0.000000,0.000000,70.545000,605.610000
50%,46.000000,0.000000,93518.000000,36.205465,-119.595293,0.000000,29.000000,25.690000,21.000000,70.050000,1394.550000,0.000000,0.000000,401.440000,2108.640000
75%,60.000000,0.000000,95329.000000,38.161321,-117.969795,3.000000,55.000000,37.680000,30.000000,89.750000,3786.600000,0.000000,0.000000,1191.100000,4801.145000
max,80.000000,9.000000,96150.000000,41.962127,-114.192901,11.000000,72.000000,49.990000,85.000000,118.750000,8684.800000,49.790000,150.000000,3564.720000,11979.340000


## Data Type Validation

Review the assigned data types and convert columns to types that more accurately represent their values and intended use.

In [8]:
# Review the distinct values of columns intended for the category data type to verify that their values are appropriate for categorical conversion.
category_columns = [
    'Gender', 
    'Offer', 
    'Internet Type',
    'Contract', 
    'Payment Method', 
    'Customer Status', 
    'Churn Category', 
    'Churn Reason'
]


for column in category_columns:
    print(churn_df[column].value_counts(dropna=False), '\n\n*===============*')

Gender
Male      3555
Female    3488
Name: count, dtype: int64 

*===============*
Offer
NaN        3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: count, dtype: int64 

*===============*
Internet Type
Fiber Optic    3035
DSL            1652
NaN            1526
Cable           830
Name: count, dtype: int64 

*===============*
Contract
Month-to-Month    3610
Two Year          1883
One Year          1550
Name: count, dtype: int64 

*===============*
Payment Method
Bank Withdrawal    3909
Credit Card        2749
Mailed Check        385
Name: count, dtype: int64 

*===============*
Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64 

*===============*
Churn Category
NaN                5174
Competitor          841
Dissatisfaction     321
Attitude            314
Price               211
Other               182
Name: count, dtype: int64 

*===============*
Churn Reason
NaN                                         

In [9]:
# Review the distinct values of columns intended for the boolean data type to verify that their values are appropriate for boolean conversion.
boolean_columns = [
    'Married',
    'Phone Service',
    'Multiple Lines',
    'Internet Service',
    'Online Security',
    'Online Backup',
    'Device Protection Plan',
    'Premium Tech Support',
    'Streaming TV',
    'Streaming Movies',
    'Streaming Music',
    'Unlimited Data',
    'Paperless Billing'
]

for column in boolean_columns:
    print(churn_df[column].value_counts(dropna=False), '\n\n*===============*')

Married
No     3641
Yes    3402
Name: count, dtype: int64 

*===============*
Phone Service
Yes    6361
No      682
Name: count, dtype: int64 

*===============*
Multiple Lines
No     3390
Yes    2971
NaN     682
Name: count, dtype: int64 

*===============*
Internet Service
Yes    5517
No     1526
Name: count, dtype: int64 

*===============*
Online Security
No     3498
Yes    2019
NaN    1526
Name: count, dtype: int64 

*===============*
Online Backup
No     3088
Yes    2429
NaN    1526
Name: count, dtype: int64 

*===============*
Device Protection Plan
No     3095
Yes    2422
NaN    1526
Name: count, dtype: int64 

*===============*
Premium Tech Support
No     3473
Yes    2044
NaN    1526
Name: count, dtype: int64 

*===============*
Streaming TV
No     2810
Yes    2707
NaN    1526
Name: count, dtype: int64 

*===============*
Streaming Movies
No     2785
Yes    2732
NaN    1526
Name: count, dtype: int64 

*===============*
Streaming Music
No     3029
Yes    2488
NaN    1526
Name: 

In [10]:
bool_mapping = {
    'Yes': True,
    'No': False
}

for column in boolean_columns:
    churn_df[column] = churn_df[column].map(bool_mapping)

In [11]:
# Group columns by their intended data type and convert
int8_columns = [
    'Age',
    'Number of Dependents',
    'Number of Referrals',
    'Tenure in Months'
]

float32_columns = [
    'Latitude',
    'Longitude',
    'Avg Monthly Long Distance Charges',
    'Avg Monthly GB Download',
    'Monthly Charge',
    'Total Charges',
    'Total Refunds',
    'Total Long Distance Charges',
    'Total Revenue'
]


dtype_mapping = {
    **dict.fromkeys(int8_columns, 'Int8'),
    **dict.fromkeys(float32_columns, 'float32'),
    **dict.fromkeys(category_columns, 'category'),
    **dict.fromkeys(boolean_columns, 'boolean'),
    'Zip Code': 'string',
    'Total Extra Data Charges': 'Int16'
}

churn_df = churn_df.astype(dtype_mapping)

In [12]:
# Review the resulting data types to confirm that the conversions were applied successfully
churn_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 38 columns):
 #   Column                             Non-Null Count  Dtype   
---  ------                             --------------  -----   
 0   Customer ID                        7043 non-null   str     
 1   Gender                             7043 non-null   category
 2   Age                                7043 non-null   Int8    
 3   Married                            7043 non-null   boolean 
 4   Number of Dependents               7043 non-null   Int8    
 5   City                               7043 non-null   str     
 6   Zip Code                           7043 non-null   string  
 7   Latitude                           7043 non-null   float32 
 8   Longitude                          7043 non-null   float32 
 9   Number of Referrals                7043 non-null   Int8    
 10  Tenure in Months                   7043 non-null   Int8    
 11  Offer                              3166 non-null   cat

## Null Values Assessment

Identify columns containing missing values and determine whether the nulls represent data quality issues or expected structural values.

In [13]:
churn_df.isna().sum()

Customer ID                             0
Gender                                  0
Age                                     0
Married                                 0
Number of Dependents                    0
City                                    0
Zip Code                                0
Latitude                                0
Longitude                               0
Number of Referrals                     0
Tenure in Months                        0
Offer                                3877
Phone Service                           0
Avg Monthly Long Distance Charges     682
Multiple Lines                        682
Internet Service                        0
Internet Type                        1526
Avg Monthly GB Download              1526
Online Security                      1526
Online Backup                        1526
Device Protection Plan               1526
Premium Tech Support                 1526
Streaming TV                         1526
Streaming Movies                  

### Offer

According to the telecom data dictionary, `Offer` represents the last marketing offer accepted by the customer. Valid values are `None` and `Offer A` through `Offer E`.

Initial inspection shows that the dataset contains `Offer A` through `Offer E` and null values, but no explicit `None` values.

In [14]:
# Check for explicit 'None' values in the Offer column
churn_df[churn_df['Offer'] == 'None']

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [15]:
# Add 'None' as a category in the Offer column and set null values in the column to it
churn_df['Offer'] = (
    churn_df['Offer']
    .cat.add_categories('None')
    .fillna('None')
)

churn_df['Offer'].value_counts(dropna=False)

Offer
None       3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: count, dtype: int64

No explicit `None` values were found. Based on the data dictionary and the observed values, the null values are interpreted as customers who did not accept a marketing offer and are replaced with `None` to make this meaning explicit.

The `Avg Monthly Long Distance Charges` and `Multiple Lines` columns contain the same number of null values. Since both fields depend on whether a customer has phone service, the null values may correspond to customers where `Phone Service` is `False`, indicating that they did not subscribe to the telco's phone service. Further investigation is needed to verify this relationship.

In [16]:
# Check for customers with 0 average monthly long-distance charges
churn_df[churn_df['Avg Monthly Long Distance Charges'] == 0]

,Customer ID,Gender,Age,Married,Number of Dependents,City,Zip Code,Latitude,Longitude,Number of Referrals,Tenure in Months,Offer,Phone Service,Avg Monthly Long Distance Charges,Multiple Lines,Internet Service,Internet Type,Avg Monthly GB Download,Online Security,Online Backup,Device Protection Plan,Premium Tech Support,Streaming TV,Streaming Movies,Streaming Music,Unlimited Data,Contract,Paperless Billing,Payment Method,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue,Customer Status,Churn Category,Churn Reason


In [17]:
churn_df['Phone Service'].value_counts(dropna=False)

Phone Service
True     6361
False     682
Name: count, dtype: Int64

In [18]:
churn_df['Multiple Lines'].value_counts(dropna=False)

Multiple Lines
False    3390
True     2971
<NA>      682
Name: count, dtype: Int64

In [19]:
# Since there are no 0 values in 'Avg Monthly Long Distance Charges', investigate whether its null values correspond to customers without phone service
# and whose 'Multiple Lines' value is also null

churn_df[(churn_df['Phone Service'] == False) & 
         (churn_df['Avg Monthly Long Distance Charges'].isnull()) & 
         (churn_df['Multiple Lines'].isnull())].shape[0]


682

In [20]:
# Impute structural nulls of 'Avg Monthly Long Distance Charges' and 'Multiple Lines' to explicitly represent customers without phone service

churn_df = churn_df.fillna(
    {
        'Avg Monthly Long Distance Charges': 0,
        'Multiple Lines': False
    }
)

In [21]:
print(churn_df['Avg Monthly Long Distance Charges'].isna().sum())
print(churn_df['Multiple Lines'].isna().sum())

0
0


### Internet Service-Related Missing Values

The following columns contain the same number of null values: `Internet Type`, `Avg Monthly GB Download`, `Online Security`, `Online Backup`, `Device Protection Plan`, `Premium Tech Support`, `Streaming TV`, `Streaming Movies`, `Streaming Music`, and `Unlimited Data`.

Since these fields are dependent on whether a customer has internet service, the null values are investigated to determine whether they correspond to customers without an internet service subscription.

In [22]:
# Check whether all internet-related null values correspond to customers without internet service
conditions = (
    churn_df['Internet Service'] == False &
    churn_df['Internet Type'].isna() &
    churn_df['Avg Monthly GB Download'].isna() &
    churn_df['Online Security'].isna() &
    churn_df['Online Backup'].isna() &
    churn_df['Device Protection Plan'].isna() &
    churn_df['Premium Tech Support'].isna() &
    churn_df['Streaming TV'].isna() &
    churn_df['Streaming Movies'].isna() &
    churn_df['Streaming Music'].isna() &
    churn_df['Unlimited Data'].isna()
)

churn_df[conditions].shape[0]


1526

In [23]:
# Replace structural nulls with values representing the absence of internet-related services

churn_df['Internet Type'] = churn_df['Internet Type'].cat.add_categories('None')

churn_df = churn_df.fillna(
    {
        'Internet Type': 'None',
        'Avg Monthly GB Download': 0,
        'Online Security': False,
        'Online Backup': False,
        'Device Protection Plan': False,
        'Premium Tech Support': False,
        'Streaming TV': False,
        'Streaming Movies': False,
        'Streaming Music': False,
        'Unlimited Data': False

    }
)

### Churn-Related Missing Values

`Churn Category` and `Churn Reason` contain the same number of null values. Since these fields describe why a customer churned, their availability is expected to depend on `Customer Status`.


In [24]:
# Investigate null values to determine whether they correspond to customers who have not churned

conditions = (
    (churn_df['Customer Status'] != 'Churned') &
    (churn_df['Churn Category'].isna()) &
    (churn_df['Churn Reason'].isna())
)

churn_df[conditions].shape[0]

5174

The null values in `Churn Category` and `Churn Reason` correspond to customers who have not churned. Since these fields are not applicable to non-churned customers, the null values are retained rather than imputed to avoid introducing artificial or misleading information into the dataset.

## Duplicate record validation

In [25]:
churn_df.duplicated().sum()

np.int64(0)

In [26]:
# Also check Customer ID specifically since each record in this dataset should have a unique Customer ID
churn_df['Customer ID'].duplicated().sum()

np.int64(0)

## Categorical Value Validation
Inspect categorical columns for unexpected values.

In [27]:
# Check category columns for inconsistencies in value naming
for column in category_columns:
    print(churn_df[column].value_counts(dropna=False))
    print('====================')

Gender
Male      3555
Female    3488
Name: count, dtype: int64
Offer
None       3877
Offer B     824
Offer E     805
Offer D     602
Offer A     520
Offer C     415
Name: count, dtype: int64
Internet Type
Fiber Optic    3035
DSL            1652
None           1526
Cable           830
Name: count, dtype: int64
Contract
Month-to-Month    3610
Two Year          1883
One Year          1550
Name: count, dtype: int64
Payment Method
Bank Withdrawal    3909
Credit Card        2749
Mailed Check        385
Name: count, dtype: int64
Customer Status
Stayed     4720
Churned    1869
Joined      454
Name: count, dtype: int64
Churn Category
NaN                5174
Competitor          841
Dissatisfaction     321
Attitude            314
Price               211
Other               182
Name: count, dtype: int64
Churn Reason
NaN                                          5174
Competitor had better devices                 313
Competitor made better offer                  311
Attitude of support person        

Compare the tenure of 'Joined' and 'Stayed' customers to verify whether the two customer statuses represent distinct tenure groups

In [28]:
churn_df[churn_df['Customer Status'] == 'Joined']['Tenure in Months'].value_counts().sort_index()

Tenure in Months
1    233
2    115
3    106
Name: count, dtype: Int64

In [29]:
churn_df[churn_df['Customer Status'] == 'Stayed']['Tenure in Months'].value_counts().sort_index()

Tenure in Months
4      93
5      69
6      70
7      80
8      81
     ... 
68     91
69     87
70    108
71    164
72    356
Name: count, Length: 69, dtype: Int64

## Numerical Value Validation
Review numerical columns for impossible, invalid, or potentially anomalous values.

In [30]:
churn_df.describe()

,Age,Number of Dependents,Latitude,Longitude,Number of Referrals,Tenure in Months,Avg Monthly Long Distance Charges,Avg Monthly GB Download,Monthly Charge,Total Charges,Total Refunds,Total Extra Data Charges,Total Long Distance Charges,Total Revenue
count,7043.0,7043.0,7043.000000,7043.000000,7043.0,7043.0,7043.000000,7043.000000,7043.000000,7043.000000,7043.000000,7043.0,7043.000000,7043.000000
mean,46.509726,0.468692,36.197456,-119.756683,1.951867,32.386767,22.958954,20.515406,63.596134,2280.381348,1.962182,6.860713,749.099243,3034.379150
std,16.750352,0.962802,2.468929,2.154425,3.001199,24.542061,15.448113,20.418941,31.204744,2266.220459,7.902614,25.104978,846.660034,2865.204590
min,19.0,0.0,32.555828,-124.301369,0.0,1.0,0.000000,0.000000,-10.000000,18.799999,0.000000,0.0,0.000000,21.360001
25%,32.0,0.0,33.990646,-121.788094,0.0,9.0,9.210000,3.000000,30.400000,400.149994,0.000000,0.0,70.544998,605.609985
50%,46.0,0.0,36.205463,-119.595291,0.0,29.0,22.889999,17.000000,70.050003,1394.550049,0.000000,0.0,401.440002,2108.639893
75%,60.0,0.0,38.161320,-117.969795,3.0,55.0,36.395000,27.000000,89.750000,3786.599976,0.000000,0.0,1191.099976,4801.145020
max,80.0,9.0,41.962128,-114.192902,11.0,72.0,49.990002,85.000000,118.750000,8684.799805,49.790001,150.0,3564.719971,11979.339844


In [31]:
# Inspect records with zero or negative 'Monthly Charge' values
churn_df.loc[
    churn_df['Monthly Charge'] <= 0,
    [
        'Customer ID',
        'Customer Status',
        'Tenure in Months',
        'Phone Service',
        'Internet Service',
        'Monthly Charge',
        'Total Charges',
        'Total Refunds'
    ]
].describe(include='all')

,Customer ID,Customer Status,Tenure in Months,Phone Service,Internet Service,Monthly Charge,Total Charges,Total Refunds
count,120,120,120.0,120,120,120.000000,120.000000,120.000000
unique,120,3,<NA>,2,2,NaN,NaN,NaN
top,0003-MKNFE,Stayed,<NA>,True,True,NaN,NaN,NaN
freq,1,84,<NA>,109,94,NaN,NaN,NaN
mean,NaN,NaN,29.766667,NaN,NaN,-5.416667,2028.856201,3.176666
std,NaN,NaN,24.643929,NaN,NaN,2.888934,2120.436035,9.855009
min,NaN,NaN,1.0,NaN,NaN,-10.000000,19.400000,0.000000
25%,NaN,NaN,7.0,NaN,NaN,-8.000000,343.025009,0.000000
50%,NaN,NaN,22.5,NaN,NaN,-5.000000,1325.975037,0.000000
75%,NaN,NaN,54.0,NaN,NaN,-3.000000,3239.512451,0.000000


In [32]:
# Examine whether negative 'Monthly Charge' values follow a consecutive or systematic pattern
churn_df.loc[
    churn_df['Monthly Charge'] <= 0,
    'Monthly Charge'
].value_counts().sort_index()

Monthly Charge
-10.0    13
-9.0      9
-8.0     12
-7.0     16
-6.0      6
-5.0     11
-4.0     17
-3.0     13
-2.0     10
-1.0     13
Name: count, dtype: int64

`Monthly Charge` contains 120 records with negative values ranging from -10 to -1. According to the data dictionary, this field represents the customer's current total monthly charge for all services. No documented explanation was found for these negative values, while the affected records otherwise contain plausible customer and billing information.

Since there is insufficient evidence to determine whether these values are erroneous or what their correct values should be, they are retained to avoid making an unsupported correction to the dataset.

In [33]:
# Export to Parquet format for use in subsequent exploratory analysis notebooks.
churn_df.to_parquet('../data/processed/telecom_customer_churn_clean.parquet', index=False)

## Zip Code Population Data Validation

Validate the supporting zip code population dataset.

In [34]:
population_df = pd.read_csv('../data/raw/telecom_zipcode_population.csv')

In [35]:
population_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1671 entries, 0 to 1670
Data columns (total 2 columns):
 #   Column      Non-Null Count  Dtype
---  ------      --------------  -----
 0   Zip Code    1671 non-null   int64
 1   Population  1671 non-null   int64
dtypes: int64(2)
memory usage: 26.2 KB


In [ ]:
# Although Zip Code contains numeric values, it is best to convert it to a string as it functions as an identifier rather than a quantitative variable.
population_df['Zip Code'] = population_df['Zip Code'].astype('string')

In [40]:
# Check for duplicates
print(population_df.duplicated().sum())
print(population_df['Zip Code'].duplicated().sum())

0
0


In [46]:
population_df['Population'].describe()

count      1671.000000
mean      20276.384201
std       20689.117300
min          11.000000
25%        1789.000000
50%       14239.000000
75%       32942.500000
max      105285.000000
Name: Population, dtype: float64

### Population Validation

The minimum population value in the dataset is 11. Since this value appears unusually low for a ZIP code, records with very low population estimates are investigated to determine whether they represent valid observations or potential data quality issues.

In [44]:
population_df[population_df['Population'] < 50]

,Zip Code,Population
60,90071,21
83,90263,11
258,91743,13
267,91759,47
432,92304,42
447,92323,40
457,92338,23
728,93435,28
1144,95140,38
1162,95225,27


Although some values were reviewed as potential anomalies, small ZIP code populations may occur depending on the geographic area and population estimation methodology.

Since the data dictionary defines these values as current population estimates and provides no reference source or year for independent validation, the original population values are retained rather than applying unsupported corrections.

In [43]:
# Check whether all customer Zip Codes have a corresponding record in the population dataset
churn_df.loc[
    ~churn_df['Zip Code'].isin(population_df['Zip Code']),
    'Zip Code'
].unique()

<ArrowStringArray>
[]
Length: 0, dtype: string

### Zip Code Population Validation Summary

The supporting ZIP code population dataset contains no missing values or duplicate ZIP codes, and all population values are positive. Unusually small population estimates were reviewed and retained because no reference source or year was provided for correction.

All ZIP codes in the customer dataset have a corresponding record in the population dataset, confirming complete referential coverage between the two datasets.

In [47]:
# Export to Parquet format for use in subsequent exploratory analysis notebooks.
population_df.to_parquet('../data/processed/telecom_zipcode_population_clean.parquet', index=False)